### ISSUE TITLE
Question about the error

### ISSUE THREAD

**Issue by bwilkerson** [sigmundch/DEP-resolved-part-of#5]
> It is a compile-time error if the specified URI does not refer to a library declaration.

Should this be stronger? I think we want to ensure that if a library L contains a part directive P that the P's part of directive refers to L.

Also, if that isn't the case we might actually want the errors to be on the library rather than in the part.

**Comment by sigmundch** [comment]
good points!

I agree we should add the check to verify that the resolved library refers back to the part.

As for where to report the error, why not both? :) - Sort of how when we report errors about duplicate definitions, dart2js often shows a hint of where is the conflicting definition defined at.

In this case, I see two scenarios for which I'd give a different error, but they present in the same way:
- someone copy/pasted a part file to use in a different library, but they forgot to update the URI: here the error is in the part
- someone copy/paste a part file to create a second part for the same library and forgot to add the part directive: here the error is in the library.

I don't think we can distinguish these scenarios, so if I had to pick, I'd use the `part` as the default location to report the issue.

**Comment by bwilkerson** [comment]
I'm fine with having two errors. One if a library includes a part that does not reference the originating library and one if a part references a library that does not include that part. (And a third error if a part references something that isn't a library).

If we only have one error, then I would argue for the library.
1. Outside an editor, people will generally analyze a library, and it makes sense to me that more errors would be reported there.
2. Inside an editor, people will reach the library first when navigating, and might never open the part to see the error.
3. It's consistent with other directive-based errors, such as an invalid or non-existent URI or importing a library whose name matches another library's name.

**Comment by sigmundch** [comment]
You make good valid points, if we go with 1 error, I agree with you now. Hopefully in that case we can provide a secondary error location to identify where the part would be fixed if the problem is in the part.

I think my vote is for having 2 errors.

**Comment by bwilkerson** [comment]
Mine too.

**Comment by lrhn** [comment]
I'm also slightly concerned about the wording "does not refer to a library declaration."
I would usually reserve the phrase "library declaration" for the actual `library ...` declaration in a library file. 
I would just write "does not refer to a Dart library file".

**Comment by bwilkerson** [comment]
I agree that's better wording. (Although I usually refer to `library ...` as a 'directive', not a 'declaration', so for me there was no confusion.)

**Comment by lrhn** [comment]
Checking it, the specification uses "library declaration" about both - your text matches the existing text for similar cases, so I guess you should keep it for consistency - but it also uses "library declaration" about the actual `library xyz` syntax at one point. The syntax production is named `libraryName`, and that section doesn't give the syntax any other name.
